ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,646 @@
+				info!(target: "engine",
+				       "Back-filling block state. last_checkpoint_number: {}, target: {}({}).",
+				       last_checkpoint_number, header.number(), header.hash());
+
+				// Get the state for last checkpoint.
+				let last_checkpoint_hash = *(chain.front().expect("Should exists").parent_hash());
+				let last_checkpoint_header = match c.block_header(BlockId::Hash(last_checkpoint_hash)) {
+					None => return Err(From::from("Unable to find last checkpoint block")),
+					Some(header) => header.decode()?,
+				};
+
+				let last_checkpoint_state: CliqueBlockState;
+
+				// We probably don't have it cached, but try anyway.
+				if let Some(st) = block_state_by_hash.get_mut(&last_checkpoint_hash) {
+					last_checkpoint_state = (*st).clone();
+				} else {
+					last_checkpoint_state = self.new_checkpoint_state(&last_checkpoint_header)?;
+				}
+				block_state_by_hash.insert(last_checkpoint_header.hash().clone(),last_checkpoint_state.clone());
+
+				// Backfill!
+				let mut new_state = last_checkpoint_state.clone();
+				for item in chain {
+					new_state.apply(item, false)?;
+				}
+				new_state.calc_next_timestamp(header, self.period);
+				block_state_by_hash.insert(header.hash(), new_state.clone());
+
+				let elapsed = backfill_start.elapsed();
+				info!(target: "engine",
+				      "Back-filling succeed, took {} ms.",
+				      elapsed.as_secs() * 1000 + elapsed.subsec_millis() as u64,
+				);
+
+				Ok(new_state)
+			}
+		}
+	}
+}
+
+impl Engine<EthereumMachine> for Clique {
+	fn name(&self) -> &str { "Clique" }
+
+	fn machine(&self) -> &EthereumMachine { &self.machine }
+	/// Clique use same fields, nonce + mixHash
+	fn seal_fields(&self, _header: &Header) -> usize { 2 }
+	fn maximum_uncle_count(&self, _block: BlockNumber) -> usize { 0 }
+
+	// No Uncle in Clique
+	fn maximum_uncle_age(&self) -> usize { 0 }
+
+	fn on_new_block(
+		&self,
+		_block: &mut ExecutedBlock,
+		_epoch_begin: bool,
+		_ancestry: &mut Iterator<Item=ExtendedHeader>,
+	) -> Result<(), Error> {
+		Ok(())
+	}
+
+	fn on_close_block(&self, _block: &mut ExecutedBlock) -> Result<(), Error> {
+		// Clique has no block reward.
+		Ok(())
+	}
+
+	fn on_seal_block(&self, block: &ExecutedBlock) -> Result<Option<Header>, Error> {
+		trace!(target: "engine", "on_seal_block");
+
+		let mut header = block.header().clone();
+
+		let state = self.state_no_backfill(header.parent_hash()).ok_or_else(
+			|| format!("on_seal_block: Unable to get parent state: {}", header.parent_hash())
+		)?;
+
+		let is_checkpoint = header.number() % self.epoch_length == 0;
+
+		header.set_author(NULL_AUTHOR);
+
+		// cast an random Vote if not checkpoint
+		if !is_checkpoint {
+			let votes = self.proposals.read().iter()
+				.filter(|item| state.is_valid_vote(item.0, *item.1))
+				.map(|item| (*item.0, *item.1))
+				.collect_vec();
+
+			if !votes.is_empty() {
+				// Pick an random vote.
+				let i = rand::thread_rng().gen_range(0 as usize, votes.len());
```

Comment:
I'd use something more descriptive here. At first glance I thought I was inside a loop.

```suggestion
				let random_vote = rand::thread_rng().gen_range(0 as usize, votes.len());
```